Mouse events should be fired when window moves out from underneath the cursor

danae404 opened the pull request:
#### d208ef00a2289c1f7c4d675a8f11f191ff458bb7
<pre>
Mouse events should be fired when window moves out from underneath the cursor
<a href="https://bugs.webkit.org/show_bug.cgi?id=304361">https://bugs.webkit.org/show_bug.cgi?id=304361</a>
<a href="https://rdar.apple.com/161493924">rdar://161493924</a>

Reviewed by Lily Spiniolas.

When the window moves and the cursor is stationary, such as when entering/exiting
fullscreen, we do not dispatch mouse events. tv.youtube.com and nba.com add mouse
event listeners to their fullscreen button, inside which they show and
hide a text description of the button. When they enter fullscreen, the button moves
out from under the cursor, but we do not dispatch a mouseleave or mouseout event,
so the label remains showing.

To fix this, when the window changes, WebPageProxy should pass to EventHandler via
EventHandler::setLastKnownMousePosition the new mouse position, and then call
EventHandler::updateMouseEventTargetAfterLayout to trigger a mouse event.

Will add test in later patch.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::setLastKnownMousePosition):
* Source/WebCore/page/EventHandler.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::windowAndViewFramesChanged):
(WebKit::WebPageProxy::updateMouseEventTargetAfterWindowAndViewFramesChanged):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::windowAndViewFramesChanged):
(WebKit::WebPageProxy::updateMouseEventTargetAfterWindowAndViewFramesChanged):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setLastKnownMousePosition):
(WebKit::WebPage::windowAndViewFramesChanged):
(WebKit::WebPage::updateMouseEventTargetAfterWindowAndViewFramesChanged):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/305529@main">https://commits.webkit.org/305529@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e03004451a5a1549069fe3ea3d50a4d943b02484

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/138254 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/10619 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/49664 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/146333 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/91229 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4cd286b7-07d3-43e6-9111-c56db1066319) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/140128 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/11323 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10773 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/105759 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/77144 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/cb0003c4-4603-4f62-b1fe-8a5e8b5b9e14) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/141201 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/8481 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/123945 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/86606 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/dc132418-e8c5-45eb-8d04-8cf7ae4095c5) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/8073 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5834 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/6615 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/117485 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/42144 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/149042 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/10301 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42701 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/114165 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/10318 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8703 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/114505 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/8040 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/120231 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/65124 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21367 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/10348 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/38167 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/10079 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/73915 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/10288 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/10139 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -209,9 +209,27 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+    RefPtr pageClient = this->pageClient();
+    if (!pageClient)
+        return;
+    RetainPtr<NSWindow> window = pageClient->platformWindow();
```
```suggestion
    RetainPtr window = pageClient->platformWindow();
```

lilyspiniolas commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -209,9 +209,27 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+    RetainPtr<NSWindow> window = pageClient->platformWindow();
+
+    NSPoint mouseLocationInWindow = [window mouseLocationOutsideOfEventStream];
+    IntPoint mousePositionInView = IntPoint(IntPoint(mouseLocationInWindow) - roundedIntPoint(viewFrameInWindowCoordinates.location()));
```
```suggestion
    IntPoint mousePositionInView = IntPoint(mouseLocationInWindow) - roundedIntPoint(viewFrameInWindowCoordinates.location());
```

lilyspiniolas commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -209,9 +209,27 @@ static inline bool expectsLegacyImplicitRubberBandControl()
     });
 }
 
+void WebPageProxy::updateMouseEventTargetAfterLayout(const FloatRect& viewFrameInWindowCoordinates)
```
This should be renamed to `updateMouseEventTargetAfterWindowAndViewFramesChanged` or something similar

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -2277,7 +2277,7 @@ class WebPage final : public API::ObjectImpl<API::Object::Type::BundlePage>, pub
     void keyEvent(WebCore::FrameIdentifier, const WebKeyboardEvent&);
 
-    void setLastKnownMousePosition(WebCore::FrameIdentifier, WebCore::IntPoint eventPoint, WebCore::IntPoint globalPoint);
+    void setLastKnownMousePosition(WebCore::FrameIdentifier, WebCore::DoublePoint eventPoint, WebCore::DoublePoint globalPoint);
```
`const WebCore::DoublePoint&` since these are twice as big as IntPoints.

smfr commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -208,10 +208,32 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+    // the point to the web coordinate system which has an origin of the top left corner.
+    NSPoint macMouseLocationInWindow = [window mouseLocationOutsideOfEventStream];
+
+    DoublePoint webMouseLocationInWindow = DoublePoint(macMouseLocationInWindow.x, [window frame].size.height - macMouseLocationInWindow.y);
```
`auto webMouseLocationInWindow`

smfr commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -208,10 +208,32 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+
+    // do same conversion as above
+    NSPoint macMouseLocationInScreen = [NSEvent mouseLocation];
+    DoublePoint webMouseLocationInScreen = DoublePoint(macMouseLocationInScreen.x, [[NSScreen mainScreen] frame].size.height - macMouseLocationInScreen.y);
```
Ditto

smfr commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -208,10 +208,32 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+
+    // do same conversion as above
+    auto macMouseLocationInScreen = [NSEvent mouseLocation];
+    auto webMouseLocationInScreen = DoublePoint(macMouseLocationInScreen.x, [[NSScreen mainScreen] frame].size.height - macMouseLocationInScreen.y);
```
This should do something like `[window convertPointFromScreen:[NSEvent mouseLocation]]`

smfr commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -208,10 +208,32 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+    // the point to the web coordinate system which has an origin of the top left corner.
+    auto macMouseLocationInWindow = [window mouseLocationOutsideOfEventStream];
+
+    auto webMouseLocationInWindow = DoublePoint(macMouseLocationInWindow.x, [window frame].size.height - macMouseLocationInWindow.y);
```
Rather than this manual Y-flip, it would be better to use one of the functions on `NSWindow` instead; take a look at what we do in other places we're converted into window content coordinates.

danae404 committed:
Mouse events should be fired when window moves out from underneath the cursor
https://bugs.webkit.org/show_bug.cgi?id=304361
rdar://161493924

Reviewed by Lily Spiniolas.

When the window moves and the cursor is stationary, such as when entering/exiting
fullscreen, we do not dispatch mouse events. tv.youtube.com and nba.com add mouse
event listeners to their fullscreen button, inside which they show and
hide a text description of the button. When they enter fullscreen, the button moves
out from under the cursor, but we do not dispatch a mouseleave or mouseout event,
so the label remains showing.

To fix this, when the window changes, WebPageProxy should pass to EventHandler via
EventHandler::setLastKnownMousePosition the new mouse position, and then call
EventHandler::updateMouseEventTargetAfterLayout to trigger a mouse event.

Will add test in later patch.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::setLastKnownMousePosition):
* Source/WebCore/page/EventHandler.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::windowAndViewFramesChanged):
(WebKit::WebPageProxy::updateMouseEventTargetAfterWindowAndViewFramesChanged):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::windowAndViewFramesChanged):
(WebKit::WebPageProxy::updateMouseEventTargetAfterWindowAndViewFramesChanged):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setLastKnownMousePosition):
(WebKit::WebPage::windowAndViewFramesChanged):
(WebKit::WebPage::updateMouseEventTargetAfterWindowAndViewFramesChanged):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/305529@main
